Index: LinqToTwitter/LinqToTwitter/TwitterContext.cs
===================================================================
--- LinqToTwitter/LinqToTwitter/TwitterContext.cs	(Revision 67699)
+++ LinqToTwitter/LinqToTwitter/TwitterContext.cs	(Working Copy)
@@ -65,7 +65,7 @@
         /// <param name="searchUrl">Base url of Twitter Search API.  May be null to use the default "http://search.twitter.com/" value.</param>
         public TwitterContext(ITwitterExecute execute, string baseUrl, string searchUrl)
         {
-            TwitterExecutor = execute ?? new TwitterExecute();
+            TwitterExecutor = execute;
             BaseUrl = string.IsNullOrEmpty(baseUrl) ? "https://api.twitter.com/1/" : baseUrl;
             SearchUrl = string.IsNullOrEmpty(searchUrl) ? "http://search.twitter.com/" : searchUrl;
         }
Index: LinqToTwitter/LinqToTwitter/TwitterExecute.cs
===================================================================
--- LinqToTwitter/LinqToTwitter/TwitterExecute.cs	(Revision 67699)
+++ LinqToTwitter/LinqToTwitter/TwitterExecute.cs	(Working Copy)
@@ -114,15 +114,6 @@
         #region Initialization
 
         /// <summary>
-        /// Initializes a new instance of the <see cref="TwitterExecute"/> class.
-        /// </summary>
-        public TwitterExecute()
-        {
-            this.AuthorizedClient = new UsernamePasswordAuthorization();
-            this.AuthorizedClient.UserAgent = m_linqToTwitterVersion;
-        }
-
-        /// <summary>
         /// supports testing
         /// </summary>
         /// <param name="oAuthTwitter">IOAuthTwitter Mock</param>
Index: LinqToTwitter/OAuth/DesktopOAuthAuthorization.cs
===================================================================
--- LinqToTwitter/OAuth/DesktopOAuthAuthorization.cs	(Revision 67699)
+++ LinqToTwitter/OAuth/DesktopOAuthAuthorization.cs	(Working Copy)
@@ -38,8 +38,8 @@
         /// <summary>
         /// Initializes a new instance of the <see cref="DesktopOAuthAuthorization"/> class.
         /// </summary>
-        public DesktopOAuthAuthorization()
-            : this(TwitterServiceDescription)
+        public DesktopOAuthAuthorization(string consumerKey, string consumerSecret)
+            : this(TwitterServiceDescription, consumerKey, consumerSecret)
         {
         }
 
@@ -70,8 +70,8 @@
         /// Initializes a new instance of the <see cref="DesktopOAuthAuthorization"/> class.
         /// </summary>
         /// <param name="serviceDescription">The service description.</param>
-        public DesktopOAuthAuthorization(ServiceProviderDescription serviceProviderDescription)
-            : base(new DesktopConsumer(serviceProviderDescription, new WindowsCredentialStoreTokenManager()))
+        public DesktopOAuthAuthorization(ServiceProviderDescription serviceProviderDescription, string consumerKey, string consumerSecret)
+            : base(new DesktopConsumer(serviceProviderDescription, new WindowsCredentialStoreTokenManager(consumerKey, consumerSecret)))
         {
             var inMemoryTokenManager = this.Consumer.TokenManager as WindowsCredentialStoreTokenManager;
             if (inMemoryTokenManager != null)
Index: LinqToTwitter/OAuth/OAuthAuthorization.cs
===================================================================
--- LinqToTwitter/OAuth/OAuthAuthorization.cs	(Revision 67699)
+++ LinqToTwitter/OAuth/OAuthAuthorization.cs	(Working Copy)
@@ -30,6 +30,12 @@
     [Serializable]
     public abstract class OAuthAuthorization : ITwitterAuthorization, IDisposable
     {
+        public IWebProxy Proxy
+        {
+            get;
+            set;
+        }
+
         /// <summary>
         /// The description of Twitter's OAuth protocol URIs.
         /// </summary>
@@ -204,6 +210,7 @@
         {
             var req = this.Consumer.PrepareAuthorizedRequest(new MessageReceivingEndpoint(requestUrl, HttpDeliveryMethods.GetRequest | HttpDeliveryMethods.AuthorizationHeaderRequest), this.AccessToken);
             this.InitializeRequest(req);
+            req.Proxy = this.Proxy;
             return req;
         }
 
@@ -216,8 +223,9 @@
         /// </returns>
         public HttpWebRequest Post(Uri requestUrl)
         {
-            var req = this.Consumer.PrepareAuthorizedRequest(new MessageReceivingEndpoint(requestUrl, HttpDeliveryMethods.PostRequest | HttpDeliveryMethods.AuthorizationHeaderRequest), this.AccessToken);
+            HttpWebRequest req = this.Consumer.PrepareAuthorizedRequest(new MessageReceivingEndpoint(requestUrl, HttpDeliveryMethods.GetRequest | HttpDeliveryMethods.AuthorizationHeaderRequest), this.AccessToken);
             this.InitializeRequest(req);
+            req.Proxy = this.Proxy;
             return req;
         }
 
@@ -234,6 +242,7 @@
             servicePoint.Expect100Continue = false;
             var req = this.Consumer.PrepareAuthorizedRequest(new MessageReceivingEndpoint(requestUrl, HttpDeliveryMethods.PostRequest), this.AccessToken);
             this.InitializeRequest(req);
+            req.Proxy = this.Proxy;
             return (HttpWebResponse)req.GetResponse();
         }
         #endregion
Index: LinqToTwitter/OAuth/WindowsCredentialStoreTokenManager.cs
===================================================================
--- LinqToTwitter/OAuth/WindowsCredentialStoreTokenManager.cs	(Revision 67699)
+++ LinqToTwitter/OAuth/WindowsCredentialStoreTokenManager.cs	(Working Copy)
@@ -59,10 +59,10 @@
         /// <summary>
         /// Initializes a new instance of the <see cref="WindowsCredentialStoreTokenManager"/> class.
         /// </summary>
-        internal WindowsCredentialStoreTokenManager()
+        internal WindowsCredentialStoreTokenManager(string consumerKey, string consumerSecret)
         {
-            this.ConsumerKey = ConfigurationManager.AppSettings["twitterConsumerKey"];
-            this.ConsumerSecret = ConfigurationManager.AppSettings["twitterConsumerSecret"];
+            this.ConsumerKey = consumerKey;
+            this.ConsumerSecret = consumerSecret;
         }
 
         /// <summary>
